pkg/helios/cas/cas.go

```diff
@@ -0,0 +1,398 @@
+	memoryMode  bool              // Skip disk I/O for maximum performance
+	writeQueue  chan writeOp      // Async write queue for background persistence
+	errorQueue  chan error        // Channel for background write errors
+	wg          sync.WaitGroup    // Wait group for background writes
+	closed      bool              // Track if store is closed
+	closeMutex  sync.Mutex        // Protects close operations
+}
+
+// NewBLAKE3Store creates a new BLAKE3-based content-addressable store
+// storePath: directory for persistent storage
+func NewBLAKE3Store(storePath string) (*BLAKE3Store, error) {
+	// Ensure storage directory exists
+	if err := os.MkdirAll(storePath, 0755); err != nil {
+		return nil, fmt.Errorf("failed to create storage directory: %w", err)
+	}
+
+	store := &BLAKE3Store{
+		storePath:  storePath,
+		cache:      make(map[string][]byte),
+		keyCache:   make(map[string]string),
+		memoryMode: false, // Default to persistent mode
+		writeQueue: make(chan writeOp, 1000), // Buffered channel for async writes
+		errorQueue: make(chan error, 100),    // Buffered channel for errors
+	}
+	
+	// Initialize hasher pool for zero-allocation hashing
+	store.hasherPool = sync.Pool{
+		New: func() interface{} {
+			return blake3.New(32, nil) // Pre-configured 256-bit unkeyed hasher
+		},
+	}
+	
+	// Start background writer goroutine
+	go store.backgroundWriter()
+	
+	// Start error handler goroutine
+	go store.errorHandler()
+	
+	return store, nil
+}
+
+// backgroundWriter handles async disk writes for maximum performance
+func (s *BLAKE3Store) backgroundWriter() {
+	for writeOp := range s.writeQueue {
+		if err := os.WriteFile(writeOp.filePath, writeOp.content, 0644); err != nil {
+			// Send error to error handler for logging/retry
+			select {
+			case s.errorQueue <- fmt.Errorf("background write failed for %s: %w", writeOp.filePath, err):
+			default:
+				// Error queue full, error will be dropped but write continues
+			}
+		}
+		s.wg.Done()
+	}
+}
+
+// errorHandler processes background write errors
+func (s *BLAKE3Store) errorHandler() {
+	for err := range s.errorQueue {
+		// Log error to stderr for proper error handling
+		// Using fmt.Fprintf to stderr instead of stdout for errors
+		fmt.Fprintf(os.Stderr, "[ERROR] BLAKE3Store background write failed: %v\n", err)
+		
+		// TODO: In production, this should:
+		// 1. Log to proper logging system (slog, logrus, zap)
+		// 2. Implement retry logic with exponential backoff
+		// 3. Send metrics/alerts for monitoring
+		// 4. Consider circuit breaker pattern for persistent failures
+	}
+}
+
+// EnableMemoryMode switches to memory-only mode for maximum performance (<70μs targets)
+func (s *BLAKE3Store) EnableMemoryMode() {
+	s.memoryMode = true
+}
+
+// Store saves content and returns its BLAKE3 hash
+func (s *BLAKE3Store) Store(content []byte) (types.Hash, error) {
+	// Quick check if store is closed (avoid lock for performance)
+	if s.closed {
```

Comment:
Potential data race and shutdown race: s.closed is read without synchronization while Close() writes it under closeMutex. Under -race this will trigger a data race. Also a caller can pass this check and still enqueue after Close begins, risking a send-on-closed panic. Suggestions:\n- Make closed an atomic flag or guard reads with the same lock.\n- Gate enqueue with a done channel: select { case s.writeQueue <- op: case <-s.done: return error or do sync write }.\n- In Close(): close(done) first, then wg.Wait(), then close queues.